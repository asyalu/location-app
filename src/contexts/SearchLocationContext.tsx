import { createContext, FC, ReactNode, useContext, useState } from 'react';
import { IIpStackResponse } from '../api/loader';
import { ISearchItem } from '../interfaces/searchItem';

interface SearchLocationContextType {
  searchLocation: IIpStackResponse | null;
  setUserLocation: (data: IIpStackResponse) => void;
  searchHistory: ISearchItem[];
  setSearchHistory: (history: ISearchItem[]) => void;
}

const SearchLocationContext = createContext<SearchLocationContextType>({
  searchLocation: null,
  setUserLocation: () => {},
  searchHistory: [],
  setSearchHistory: () => {},
});

const useSearchLocationContext = () => useContext(SearchLocationContext);

const SearchLocationProvider: FC<{ children: ReactNode }> = ({ children }) => {
  const [searchLocation, setUserLocation] = useState<IIpStackResponse | null>(
    null
  );
  const [searchHistory, setSearchHistory] = useState<ISearchItem[]>([]);

  return (
    <SearchLocationContext.Provider
      value={{
        searchLocation,
        setUserLocation,
        searchHistory,
        setSearchHistory,
      }}
    >
      {children}
    </SearchLocationContext.Provider>
  );
};

export { useSearchLocationContext, SearchLocationProvider };
