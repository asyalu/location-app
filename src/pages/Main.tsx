const Main = () => <div>start</div>;

export default Main;
