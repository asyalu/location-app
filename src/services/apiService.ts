import { getLocation, IIpStackResponse } from '../api/loader';

const fetchLocationData = async (
  searchValue?: string
): Promise<IIpStackResponse> => {
  try {
    const response = await getLocation(searchValue);
    return response;
  } catch (error) {
    throw new Error('Error fetching location data');
  }
};

export { fetchLocationData };
