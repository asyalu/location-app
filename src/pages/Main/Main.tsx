import { useEffect, useState } from 'react';
import { Map } from '../../components/Map/Map';
import { IIpStackResponse } from '../../api/loader';
import { ListOfSearches } from '../../components/ListOfSearches/ListOfSearches';
import { LocationInfo } from '../../components/LocationInfo/LocationInfo';
import { SearchBox } from '../../components/SearchBox/SearchBox';
import { useSearchLocationContext } from '../../contexts/SearchLocationContext';
import { fetchLocationData } from '../../services/apiService';

const Main = () => {
  const [userLocation, setUserLocationInfo] = useState<
    IIpStackResponse | undefined
  >();
  const { searchLocation } = useSearchLocationContext();

  const getUserLocationInfo = async () => {
    const response = await fetchLocationData();

    setUserLocationInfo(response);
  };

  useEffect(() => {
    getUserLocationInfo();
  }, []);

  return (
    <div className="h-full flex p-5">
      <div className="w-1/3 mt-5 p-4 rounded-lg shadow">
        <ListOfSearches />
      </div>
      <div className="w-full flex flex-col gap-5 p-5">
        <div className="w-full flex gap-5">
          {userLocation ? (
            <>
              <div className="w-full flex justify-center bg-white p-4 rounded-lg shadow">
                <Map
                  lat={userLocation.latitude}
                  lng={userLocation.longitude}
                />
              </div>
              <div className="w-full bg-white p-4 rounded-lg shadow">
                <h2 className="text-xl font-bold mb-4">
                  Information about user location
                </h2>
                <LocationInfo
                  city={userLocation.city}
                  country={userLocation.country_name}
                  ip={userLocation.ip}
                />
              </div>
            </>
          ) : (
            <span className="text-gray-500">Loading user location...</span>
          )}
        </div>
        <div className="w-full">
          <SearchBox />
        </div>
        {searchLocation ? (
          <div className="w-full flex gap-5">
            <div className="w-full flex justify-center bg-white p-4 rounded-lg shadow">
              <Map
                lat={searchLocation.latitude}
                lng={searchLocation.longitude}
              />
            </div>
            <div className="w-full bg-white p-4 rounded-lg shadow">
              <h2 className="text-xl font-bold mb-4">
                Information about search location
              </h2>
              <LocationInfo
                city={searchLocation.city}
                country={searchLocation.country_name}
                ip={searchLocation.ip}
              />
            </div>
          </div>
        ) : (
          <span className="text-gray-500 flex items-center justify-center h-30%">
            Waiting for your search to display the data...
          </span>
        )}
      </div>
    </div>
  );
};

export { Main };
