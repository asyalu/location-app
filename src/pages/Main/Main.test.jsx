import '@testing-library/jest-dom/extend-expect';
import { render, screen, act } from '@testing-library/react';
import { Main } from './Main';
import { useSearchLocationContext } from '../../contexts/SearchLocationContext';
import { fetchLocationData } from '../../services/apiService';

jest.mock('../../contexts/SearchLocationContext', () => ({
  useSearchLocationContext: jest.fn(),
}));

jest.mock('../../services/apiService', () => ({
  fetchLocationData: jest.fn(),
}));

const mockLocationData = {
  latitude: 12.34,
  longitude: -43.21,
  city: 'London',
  country_name: 'UK',
  ip: '123.555.01.33',
};

const mockHistory = {
  id: 1,
  query: '44.6666.12.4400',
};

beforeEach(async () => {
  fetchLocationData.mockResolvedValue(mockLocationData);

  useSearchLocationContext.mockReturnValue({
    locationSearchData: [mockLocationData],
    historySearchData: [mockHistory],
  });

  await act(async () => {
    render(<Main />);
  });
});

test('matches Main component snapshot', () => {
  const { asFragment } = render(<Main />);
  expect(asFragment()).toMatchSnapshot();
});

test('renders Main component with user location data', async () => {
  const locationInfoElement = await screen.findByText(
    'Information about user location'
  );
  expect(locationInfoElement).toBeInTheDocument();
});
