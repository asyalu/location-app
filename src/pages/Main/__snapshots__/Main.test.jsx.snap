// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches Main component snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-full flex p-5"
  >
    <div
      class="w-1/3 mt-5 p-4 rounded-lg shadow"
    >
      <div
        class="p-4"
      >
        <h2
          class="text-xl font-bold mb-4 bg-gray-100 p-2 rounded-lg shadow"
        >
          List of searches
        </h2>
        <ul
          class="space-y-2 "
        >
          <li
            class="bg-white p-2 rounded-lg shadow cursor-pointer hover:bg-gray-200"
          >
            44.6666.12.4400
          </li>
        </ul>
      </div>
    </div>
    <div
      class="w-full flex flex-col gap-5 p-5"
    >
      <div
        class="w-full flex gap-5"
      >
        <span
          class="text-gray-500"
        >
          Loading user location...
        </span>
      </div>
      <div
        class="w-full"
      >
        <form
          class="w-full"
        >
          <div
            class="h-10% w-full bg-gray-100 p-4 flex items-center"
          >
            <input
              class="w-3/4 border border-gray-300 p-2 mr-2 rounded"
              placeholder="Enter IP or URL..."
              type="text"
              value=""
            />
            <button
              class="bg-blue-500 text-white px-4 py-2 rounded"
              type="submit"
            >
              Search
            </button>
          </div>
        </form>
      </div>
      <div
        class="w-full flex gap-5"
      >
        <div
          class="w-full flex justify-center bg-white p-4 rounded-lg shadow"
        >
          <span>
            Loading...
          </span>
        </div>
        <div
          class="w-full bg-white p-4 rounded-lg shadow"
        >
          <h2
            class="text-xl font-bold mb-4"
          >
            Information about search location
          </h2>
          <div
            class="h-30% w-30% bg-gray-100 p-4 rounded-lg shadow"
          >
            <p>
              <span
                class="font-bold"
              >
                City:
              </span>
               
            </p>
            <p>
              <span
                class="font-bold"
              >
                Country:
              </span>
               
            </p>
            <p>
              <span
                class="font-bold"
              >
                IP:
              </span>
               
            </p>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
